cuda_pathfinder/cuda/pathfinder/_dynamic_libs/load_dl_linux.py

```diff
@@ -14,48 +14,40 @@
 
 LIBDL_PATH = ctypes.util.find_library("dl") or "libdl.so.2"
 LIBDL = ctypes.CDLL(LIBDL_PATH)
-LIBDL.dladdr.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
-LIBDL.dladdr.restype = ctypes.c_int
+LIBDL.dlinfo.argtypes = [ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p]
+LIBDL.dlinfo.restype = ctypes.c_int
 
 
-class DlInfo(ctypes.Structure):
-    """Structure used by dladdr to return information about a loaded symbol."""
+# First appeared in 2004-era glibc. Universally correct on Linux for all practical purposes.
+RTLD_DI_LINKMAP = 2
 
+
+class LinkMap(ctypes.Structure):
+    # Minimal fields we need; layout matches glibc's struct link_map
     _fields_ = (
-        ("dli_fname", ctypes.c_char_p),  # path to .so
-        ("dli_fbase", ctypes.c_void_p),
-        ("dli_sname", ctypes.c_char_p),
-        ("dli_saddr", ctypes.c_void_p),
+        ("l_addr", ctypes.c_void_p),
+        ("l_name", ctypes.c_char_p),
+        ("l_ld", ctypes.c_void_p),
+        ("l_next", ctypes.c_void_p),
+        ("l_prev", ctypes.c_void_p),
     )
 
 
-def abs_path_for_dynamic_library(libname: str, handle: ctypes.CDLL) -> Optional[str]:
-    """Get the absolute path of a loaded dynamic library on Linux.
-
-    Args:
-        libname: The name of the library
-        handle: The library handle
+def abs_path_for_dynamic_library(libname: str, handle: ctypes.CDLL) -> str:
+    lm_ptr = ctypes.POINTER(LinkMap)()
+    rc = LIBDL.dlinfo(ctypes.c_void_p(handle._handle), RTLD_DI_LINKMAP, ctypes.byref(lm_ptr))
+    if rc == 0 and lm_ptr and lm_ptr.contents.l_name:
+        path: str = lm_ptr.contents.l_name.decode()
+        if path:
+            return path
 
-    Returns:
-        The absolute path to the library file, or None if no expected symbol is found
+    raise OSError(f"abs_path_for_dynamic_library failed for {libname=!r}")
```

Comment:
One potential improvement for error handling is we subsequently call `dlerror` to get the err msg if -1 is returned by `dladdr`/`dlinfo`.

Reply:
Done, also commit be2078207342031ddbff71679aeca032173490d9.